File: solutions/python/rotational-cipher/1/rotational_cipher.py
def rotate_letter(char: str, base_code: int, key: int) -> str:
    return chr((ord(char) - base_code + key) % 26 + base_code)


def rotate_char(char: str, key: int) -> str:
    if char.isupper():
        return rotate_letter(char, ord("A"), key)
    if char.islower():
        return rotate_letter(char, ord("a"), key)
    return char


def rotate(text: str, key: int) -> str:
    return "".join(rotate_char(char, key) for char in text)

File: solutions/c/rotational-cipher/1/rotational_cipher.c
#include "rotational_cipher.h"

#include <ctype.h>
#include <stdlib.h>
#include <string.h>

char rotate_letter(char c, char base, int key);
char rotate_letter(char c, char base, int key) {
    return (c - base + key) % 26 + base;
}

char rotate_char(char c, int key);
char rotate_char(char c, int key) {
    return isupper(c)   ? rotate_letter(c, 'A', key)
           : islower(c) ? rotate_letter(c, 'a', key)
                        : c;
}

char* rotate(const char* text, int key) {
    size_t const n = strlen(text);
    char* out = malloc(n + 1);
    for (size_t i = 0; i <= n; ++i) {
        out[i] = rotate_char(text[i], key);
    }
    return out;
}
